compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -698,10 +698,16 @@ impl<'tcx> TypeVisitor<'tcx> for OrphanChecker<'tcx> {
             }
             ty::Error(_) => ControlFlow::Break(OrphanCheckEarlyExit::LocalTy(ty)),
             ty::Closure(..) | ty::Generator(..) | ty::GeneratorWitness(..) => {
-                self.tcx.sess.delay_span_bug(
-                    DUMMY_SP,
-                    format!("ty_is_local invoked on closure or generator: {:?}", ty),
-                );
+                if !self.tcx.features().specialization {
+                    // With specialization, there could be a blanket impl `impl<T: Send> Trait for T`
+                    // as well as an `impl Trait for OpaqueType`, where the hidden type of `OpaqueType`
+                    // is a closure or generator. See tests/ui/type-alias-impl-trait/issue-104817.rs for
+                    // an example.
+                    self.tcx.sess.delay_span_bug(
+                        DUMMY_SP,
+                        format!("ty_is_local invoked on closure or generator: {:?}", ty),
+                    );
+                }
                 ControlFlow::Break(OrphanCheckEarlyExit::LocalTy(ty))
```

Comment:
in that case we should stop treating them as local types though :grin: 

probably always use `found_non_local_ty` or check the `DefId` of the item if you want :shrug:

Reply:
oh right... the closure may be from a different crate.